src/functions/ensureEnv.ts

```diff
@@ -0,0 +1,161 @@
+import * as backend from "../deploy/functions/backend";
+import * as deploymentTool from "../deploymentTool";
+import * as envstore from "./envstore";
+import * as getProjectId from "../getProjectId";
+import * as helper from "../deploy/functions/functionsDeployHelper";
+
+const ENVSTORE_INTERNAL_ID = "firebase-functions-internal";
+const CONFIGSTORE_KEY = "envstore";
+const CONFIGSTORE_TTL = 1000 * 60 * 60 * 24; // 1 day
+const DEFAULT_ENV_KEYS = ["FIREBASE_CONFIG"];
+
+/**
+ * Check if the EnvStore API is active.
+ *
+ * This is an bespoke method of checking whether user has opted in
+ * to activate the EnvStore API.
+ *
+ * We check for existance of non-emtpy ENV_ID=${ENVSTORE_INTERNAL_ID}
+ * which is created when user "enables" the EnvStore API.
+ */
+async function _check(projectId: string): Promise<boolean> {
+  const resp = await envstore.getStore(projectId, ENVSTORE_INTERNAL_ID);
+  return !!resp.vars;
+}
+
+/**
+ * Cached check for active EnvStore API.
+ */
+export async function check(projectId: string): Promise<boolean> {
+  // Check actice state from local cache.
+  const cached = configstore.get(CONFIGSTORE_KEY) as { lastActiveAt: string } | undefined;
+  if (cached?.lastActiveAt) {
+    const activeAt = new Date(cached.lastActiveAt);
+    const diff = Date.now() - activeAt.getTime();
+    if (diff <= CONFIGSTORE_TTL) {
+      return true;
+    }
+    configstore.delete(CONFIGSTORE_KEY);
+  }
+
+  // Query the EnvStore API to check active state.
+  const checked = await _check(projectId);
+  if (checked) {
+    configstore.set(CONFIGSTORE_KEY, { lastActiveAt: Date.now() });
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Enable the EnvStore API.
+ *
+ * This is an bespoke method of "enabling" EnvStore API. We "enable" the
+ * EnvStore API setting up a non-empty EnvStore ENV_ID=${ENVSTORE_INTERNAL_ID}.
+ */
+export async function enable(projectId: string): Promise<void> {
+  await envstore.patchStore(projectId, ENVSTORE_INTERNAL_ID, { ENABLED: "1" });
+}
+
+interface UserEnv {
+  fnLabel: string;
+  envs: Record<string, string>;
+}
+
+/**
+ * Lookup existing cloud functions and collect user-defined env variables.
+ *
+ * If user's project hasn't activated the EnvStore API, we consider
+ * all non-default environment variable as "user-defined". The user-defined
+ * env is significant because they will be removed on the next deploy following
+ * EnvStore API activation.
+ */
+async function getUserEnvs(projectId: string): Promise<UserEnv[]> {
+  const have = await backend.existingBackend({ projectId, filters: [] }, /* forceRefresh= */ true);
+
+  const fnEnvs: UserEnv[] = [];
+  for (const fn of have.cloudFunctions) {
+    // Filter out non CF3 function instances.
+    if (!deploymentTool.isFirebaseManaged(fn.labels || {})) {
+      continue;
+    }
+    const uenvs: Record<string, string> = {};
+    const envs = fn.environmentVariables;
+    if (envs && Object.keys(envs).length > 1) {
```

Comment:
Nit: I hate superflous blocks. You don't actually need the if block here:

```
const envs = fn.environmentVariables || {}
for (const [k, v] of Object.entries(envs)) {
  ...
}
```

Reply:
This code is a bit too subtle in that we are actually checking to see if there are non-default environment variables on the given function.

I believe there's only 1 environment variables that we consider default `FIREBASE_CONFIG`, so the code assumes that if there is more than 1 environment variable, the others must be user-defined.

I modified the code to be less verbose and hopefully more clear.